src/AspNetCoreModuleV2/CommonLib/PipeOutputManager.cpp

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+#include "stdafx.h"
+
+PipeOutputManager::PipeOutputManager() :
+    m_dwStdErrReadTotal(0),
+    m_hLogFileHandle(INVALID_HANDLE_VALUE),
+    m_hErrReadPipe(INVALID_HANDLE_VALUE),
+    m_hErrWritePipe(INVALID_HANDLE_VALUE),
+    m_hErrThread(INVALID_HANDLE_VALUE),
+    m_fDisposed(FALSE)
+{
+}
+
+PipeOutputManager::~PipeOutputManager()
+{
+    StopOutputRedirection();
+}
+
+VOID
+PipeOutputManager::StopOutputRedirection()
+{
+    DWORD    dwThreadStatus = 0;
+    STRU     struStdOutput;
+
+    if (m_fDisposed)
+    {
+        return;
+    }
+    m_fDisposed = true;
+
+    fflush(stdout);
+    fflush(stderr);
+
+    if (m_hErrWritePipe != INVALID_HANDLE_VALUE)
+    {
+        CloseHandle(m_hErrWritePipe);
+        m_hErrWritePipe = INVALID_HANDLE_VALUE;
+    }
+
+    if (m_hErrThread != NULL &&
+        GetExitCodeThread(m_hErrThread, &dwThreadStatus) != 0 &&
+        dwThreadStatus == STILL_ACTIVE)
+    {
+        // wait for gracefullshut down, i.e., the exit of the background thread or timeout
+        if (WaitForSingleObject(m_hErrThread, PIPE_OUTPUT_THREAD_TIMEOUT) != WAIT_OBJECT_0)
+        {
+            // if the thread is still running, we need kill it first before exit to avoid AV
+            if (GetExitCodeThread(m_hErrThread, &dwThreadStatus) != 0 && dwThreadStatus == STILL_ACTIVE)
+            {
+                TerminateThread(m_hErrThread, STATUS_CONTROL_C_EXIT);
+            }
+        }
+    }
+
+    CloseHandle(m_hErrThread);
+    m_hErrThread = NULL;
+
+    if (m_hErrReadPipe != INVALID_HANDLE_VALUE)
+    {
+        CloseHandle(m_hErrReadPipe);
+        m_hErrReadPipe = INVALID_HANDLE_VALUE;
+    }
+
+    _dup2(m_fdStdOut, _fileno(stdout));
+    _dup2(m_fdStdErr, _fileno(stderr));
+
+    // Write the remaining contents to the original stdout
+    if (GetStdOutContent(&struStdOutput))
+    {
+        wprintf(struStdOutput.QueryStr());
```

Comment:
this wprintf will not be redirected. Is this expected

Reply:
Yes. We want to log anything captured in the pipe to the original stdout. For example, our tests redirect stdout on iisexpress to it's own process, so we would like preserve that if possible.